compiler-core/src/language_server/messages.rs

```diff
@@ -193,7 +216,58 @@ impl MessageBuffer {
         Next::MorePlease
     }
 
-    fn response(&mut self, _: lsp_server::Response) -> Next {
+    pub fn make_request(
+        &mut self,
+        method: impl Into<String>,
+        params: impl serde::Serialize,
+        handler: Option<ResponseHandler>,
+    ) -> lsp_server::Request {
+        let id = self.next_request_id;
+        self.next_request_id += 1;
+        let request = lsp_server::Request {
+            id: id.into(),
+            method: method.into(),
+            params: serde_json::value::to_value(params).expect("serialisation should never fail"),
+        };
+
+        if let Some(handler) = handler {
+            _ = self.response_handlers.insert(id.into(), handler);
+        }
+
+        request
+    }
+
+    fn configuration_update_received(&mut self, result: serde_json::Value) -> Next {
+        let Some(first_el) = result.as_array().and_then(|a| a.first()) else {
+            return Next::MorePlease;
+        };
+
+        let parsed_config_result: Result<Configuration, _> =
+            serde_json::from_value(first_el.clone());
+
+        let Ok(parsed_config) = parsed_config_result else {
+            return Next::MorePlease;
+        };
+
+        let message = Message::Response(Response::Configuration(parsed_config));
+        self.messages.push(message);
+
+        Next::Handle(self.take_messages())
+    }
+
+    fn handle_response(&mut self, handler: ResponseHandler, result: serde_json::Value) -> Next {
+        match handler {
+            ResponseHandler::UpdateConfiguration => self.configuration_update_received(result),
+        }
+    }
+
+    fn response(&mut self, response: lsp_server::Response) -> Next {
+        if let Some(handler) = self.response_handlers.remove(&response.id) {
```

Comment:
When would it be valid for there to be a response that we don't have a handler for? Or there being a handler but no result in the response?